### ISSUE TITLE
Error: TypeError: 'NoneType' object is not iterable when reconciling gene trees and species tree

### ISSUE THREAD

**Issue by fei0810** [davidemms/OrthoFinder#310]
Hello there,

I have been encountering an error while running OrthoFinder v2.3.7 in python2 environment.

My input file are 10 pep fasta files, my command is

```
orthofinder -t 50 -a 10 -p ./ -f pepall
```

```

Inferring gene and species trees
--------------------------------
2019-11-20 10:49:17 : Done 0 of 27870
2019-11-20 10:49:20 : Done 1000 of 27870
2019-11-20 10:49:23 : Done 2000 of 27870
2019-11-20 10:49:26 : Done 3000 of 27870
2019-11-20 10:49:29 : Done 4000 of 27870
2019-11-20 10:49:32 : Done 5000 of 27870
2019-11-20 10:49:35 : Done 6000 of 27870
2019-11-20 10:49:38 : Done 7000 of 27870
2019-11-20 10:49:41 : Done 8000 of 27870
2019-11-20 10:49:44 : Done 9000 of 27870
2019-11-20 10:49:47 : Done 10000 of 27870
2019-11-20 10:49:50 : Done 11000 of 27870
2019-11-20 10:49:53 : Done 12000 of 27870
2019-11-20 10:49:56 : Done 13000 of 27870
2019-11-20 10:49:59 : Done 14000 of 27870
2019-11-20 10:50:02 : Done 15000 of 27870
2019-11-20 10:50:05 : Done 16000 of 27870
2019-11-20 10:50:08 : Done 17000 of 27870
2019-11-20 10:50:11 : Done 18000 of 27870
2019-11-20 10:50:14 : Done 19000 of 27870
2019-11-20 10:50:17 : Done 20000 of 27870
2019-11-20 10:50:20 : Done 21000 of 27870
2019-11-20 10:50:23 : Done 22000 of 27870
2019-11-20 10:50:26 : Done 23000 of 27870
2019-11-20 10:50:29 : Done 24000 of 27870
2019-11-20 10:50:32 : Done 25000 of 27870
2019-11-20 10:50:35 : Done 26000 of 27870
2019-11-20 10:50:38 : Done 27000 of 27870

OG0026960_tree_id.txt - WARNING: ETE could not interpret tree file, it will be ignored
11323 trees had all species present and will be used by STAG to infer the species tree

Best outgroup(s) for species tree
---------------------------------
2019-11-20 11:24:06 : Starting STRIDE
2019-11-20 11:24:30 : Done STRIDE
Observed 325 well-supported, non-terminal duplications. 309 support the best root and 16 contradict it.
Best outgroup for species tree:
  Triticeae_barley

Reconciling gene trees and species tree
---------------------------------------
Outgroup: Triticeae_barley
2019-11-20 11:24:30 : Starting Recon and orthologues
2019-11-20 11:24:30 : Starting OF Orthologues
2019-11-20 11:24:33 : Done 0 of 27870
2019-11-20 11:26:15 : Done 1000 of 27870
2019-11-20 11:27:04 : Done 2000 of 27870
2019-11-20 11:27:45 : Done 3000 of 27870
2019-11-20 11:28:17 : Done 4000 of 27870
2019-11-20 11:28:45 : Done 5000 of 27870
2019-11-20 11:29:09 : Done 6000 of 27870
2019-11-20 11:29:29 : Done 7000 of 27870
2019-11-20 11:29:49 : Done 8000 of 27870
2019-11-20 11:30:08 : Done 9000 of 27870
2019-11-20 11:30:25 : Done 10000 of 27870
2019-11-20 11:30:42 : Done 11000 of 27870
2019-11-20 11:30:59 : Done 12000 of 27870
2019-11-20 11:31:16 : Done 13000 of 27870
2019-11-20 11:31:33 : Done 14000 of 27870
2019-11-20 11:31:50 : Done 15000 of 27870
2019-11-20 11:32:07 : Done 16000 of 27870
2019-11-20 11:32:22 : Done 17000 of 27870
2019-11-20 11:32:37 : Done 18000 of 27870
2019-11-20 11:32:53 : Done 19000 of 27870
2019-11-20 11:33:09 : Done 20000 of 27870
2019-11-20 11:33:22 : Done 21000 of 27870
2019-11-20 11:33:36 : Done 22000 of 27870
2019-11-20 11:33:48 : Done 23000 of 27870
2019-11-20 11:34:00 : Done 24000 of 27870
2019-11-20 11:34:09 : Done 25000 of 27870
2019-11-20 11:34:18 : Done 26000 of 27870
Traceback (most recent call last):
  File "orthofinder/orthofinder.py", line 1647, in <module>
  File "orthofinder/orthofinder.py", line 1482, in GetOrthologues
  File "orthofinder/scripts/orthologues.py", line 1002, in OrthologuesWorkflow
  File "orthofinder/scripts/orthologues.py", line 800, in ReconciliationAndOrthologues
  File "orthofinder/scripts/trees2ologs_of.py", line 644, in DoOrthologuesForOrthoFinder
TypeError: 'NoneType' object is not iterable
[3921] Failed to execute script orthofinder

```

But I didn't find error in log file.

I also checked "orthofinder/scripts/trees2ologs_of.py", line 644, it is
```
species = list(speciesDict.keys())
```

I don't know what is wrong

All the best
FeiZhao

**Comment by fei0810** [comment]
Hi , I changed my input files name to XX_XX.faa format, it works now. But i find an WARNING:

```
WARNING: program called by OrthoFinder produced output to stderr

Command: ['mcl', 'OrthoFinder/Results_wheatall_ortho/WorkingDirectory/OrthoFinder_graph.txt', '-I', '1.5', '-o', 'OrthoFinder/Results_wheatall_ortho/WorkingDirectory/clusters_OrthoFinder_I1.5.txt', '-te', '10', '-V', 'all']

stdout
------

stderr
------
[mcl] cut <4> instances of overlap

2019-11-20 10:42:09 : Ran MCL

Writing orthogroups to file
---------------------------
OrthoFinder assigned 335932 genes (95.3% of total) to 35291
orthogroups. Fifty percent of all genes were in orthogroups
with 10 or more genes (G50 was 10) and were contained in the
largest 10172 orthogroups (O50 was 10172). There were 11323
orthogroups with all species present and 6291 of these
consisted entirely of single-copy genes.

2019-11-20 10:44:32 : Done orthogroups

```

**Comment by fei0810** [comment]
Update, I run the command in python3, the stdout have 2 warning messages and 1 error as below:

```
WARNING: program called by OrthoFinder produced output to stderr

Command: ['mcl', '/mypath/OrthoFinder/triticeae/OrthoFinder/Results_triticeae_ortho/WorkingDirectory/OrthoFinder_graph.txt', '-I', '1.5', '-o', '/mypath/OrthoFinder/triticeae/OrthoFinder/Results_triticeae_ortho/WorkingDirectory/clusters_OrthoFinder_I1.5.txt', '-te', '1', '-V', 'all']

stdout
------

stderr
------
[mcl] cut <4> instances of overlap

```

```
OG0026960_tree_id.txt - WARNING: ETE could not interpret tree file, it will be ignored
11323 trees had all species present and will be used by STAG to infer the species tree
```

the error 
```
Traceback (most recent call last):
  File "orthofinder/orthofinder.py", line 1647, in <module>
  File "orthofinder/orthofinder.py", line 1482, in GetOrthologues
  File "orthofinder/scripts/orthologues.py", line 1002, in OrthologuesWorkflow
  File "orthofinder/scripts/orthologues.py", line 800, in ReconciliationAndOrthologues
  File "orthofinder/scripts/trees2ologs_of.py", line 644, in DoOrthologuesForOrthoFinder
TypeError: 'NoneType' object is not iterable
[63168] Failed to execute script orthofinder
```

all stdout

```
./orthofinder -t 80 -p ./ -f triticeae -n triticeae_ortho

OrthoFinder version 2.3.7 Copyright (C) 2014 David Emms

2019-11-20 13:21:32 : Starting OrthoFinder
80 thread(s) for highly parallel tasks (BLAST searches etc.)
1 thread(s) for OrthoFinder algorithm

Checking required programs are installed
----------------------------------------
Test can run "mcl -h" - ok
Test can run "fastme -i /mypath/OrthoFinder/triticeae/OrthoFinder/Results_triticeae_ortho/WorkingDirectory/SimpleTest.phy -o /mypath/OrthoFinder/triticeae/OrthoFinder/Results_triticeae_ortho/WorkingDirectory/SimpleTest.tre" - ok

Dividing up work for BLAST for parallel processing
--------------------------------------------------
2019-11-20 13:21:42 : Creating diamond database 1 of 10
2019-11-20 13:21:42 : Creating diamond database 2 of 10
2019-11-20 13:21:43 : Creating diamond database 3 of 10
2019-11-20 13:21:43 : Creating diamond database 4 of 10
2019-11-20 13:21:44 : Creating diamond database 5 of 10
2019-11-20 13:21:44 : Creating diamond database 6 of 10
2019-11-20 13:21:45 : Creating diamond database 7 of 10
2019-11-20 13:21:45 : Creating diamond database 8 of 10
2019-11-20 13:21:46 : Creating diamond database 9 of 10
2019-11-20 13:21:47 : Creating diamond database 10 of 10

Running diamond all-versus-all
------------------------------
Using 80 thread(s)
2019-11-20 13:21:47 : This may take some time....
2019-11-20 13:21:48 : Done 0 of 100
2019-11-20 13:36:18 : Done 10 of 100
2019-11-20 13:37:20 : Done 20 of 100
2019-11-20 13:49:16 : Done all-versus-all sequence search

Running OrthoFinder algorithm
-----------------------------
2019-11-20 13:49:20 : Initial processing of each species
2019-11-20 13:52:33 : Initial processing of species 0 complete
2019-11-20 13:55:52 : Initial processing of species 1 complete
2019-11-20 13:59:38 : Initial processing of species 2 complete
2019-11-20 14:02:45 : Initial processing of species 3 complete
2019-11-20 14:05:53 : Initial processing of species 4 complete
2019-11-20 14:09:05 : Initial processing of species 5 complete
2019-11-20 14:12:15 : Initial processing of species 6 complete
2019-11-20 14:15:43 : Initial processing of species 7 complete
2019-11-20 14:19:12 : Initial processing of species 8 complete
2019-11-20 14:22:35 : Initial processing of species 9 complete
2019-11-20 14:26:15 : Connected putative homologues
2019-11-20 14:26:43 : Written final scores for species 0 to graph file
2019-11-20 14:27:23 : Written final scores for species 1 to graph file
2019-11-20 14:28:04 : Written final scores for species 2 to graph file
2019-11-20 14:28:39 : Written final scores for species 3 to graph file
2019-11-20 14:29:16 : Written final scores for species 4 to graph file
2019-11-20 14:29:47 : Written final scores for species 5 to graph file
2019-11-20 14:30:16 : Written final scores for species 6 to graph file
2019-11-20 14:30:51 : Written final scores for species 7 to graph file
2019-11-20 14:31:31 : Written final scores for species 8 to graph file
2019-11-20 14:32:02 : Written final scores for species 9 to graph file

WARNING: program called by OrthoFinder produced output to stderr

Command: ['mcl', '/mypath/OrthoFinder/triticeae/OrthoFinder/Results_triticeae_ortho/WorkingDirectory/OrthoFinder_graph.txt', '-I', '1.5', '-o', '/mypath/OrthoFinder/triticeae/OrthoFinder/Results_triticeae_ortho/WorkingDirectory/clusters_OrthoFinder_I1.5.txt', '-te', '1', '-V', 'all']

stdout
------

stderr
------
[mcl] cut <4> instances of overlap

2019-11-20 14:35:20 : Ran MCL

Writing orthogroups to file
---------------------------
OrthoFinder assigned 335932 genes (95.3% of total) to 35291
orthogroups. Fifty percent of all genes were in orthogroups
with 10 or more genes (G50 was 10) and were contained in the
largest 10172 orthogroups (O50 was 10172). There were 11323
orthogroups with all species present and 6291 of these
consisted entirely of single-copy genes.

2019-11-20 14:37:51 : Done orthogroups

Analysing Orthogroups
=====================

Calculating gene distances
--------------------------
2019-11-20 15:07:02 : Done

Inferring gene and species trees
--------------------------------
2019-11-20 15:07:04 : Done 0 of 27870
2019-11-20 15:07:07 : Done 1000 of 27870
2019-11-20 15:07:10 : Done 2000 of 27870
2019-11-20 15:07:13 : Done 3000 of 27870
2019-11-20 15:07:16 : Done 4000 of 27870
2019-11-20 15:07:19 : Done 5000 of 27870
2019-11-20 15:07:22 : Done 6000 of 27870
2019-11-20 15:07:25 : Done 7000 of 27870
2019-11-20 15:07:28 : Done 8000 of 27870
2019-11-20 15:07:31 : Done 9000 of 27870
2019-11-20 15:07:34 : Done 10000 of 27870
2019-11-20 15:07:37 : Done 11000 of 27870
2019-11-20 15:07:40 : Done 12000 of 27870
2019-11-20 15:07:43 : Done 13000 of 27870
2019-11-20 15:07:46 : Done 14000 of 27870
2019-11-20 15:07:51 : Done 15000 of 27870
2019-11-20 15:07:58 : Done 16000 of 27870
2019-11-20 15:08:01 : Done 17000 of 27870
2019-11-20 15:08:08 : Done 18000 of 27870
2019-11-20 15:08:11 : Done 19000 of 27870
2019-11-20 15:08:15 : Done 20000 of 27870
2019-11-20 15:08:18 : Done 21000 of 27870
2019-11-20 15:08:21 : Done 22000 of 27870
2019-11-20 15:08:24 : Done 23000 of 27870
2019-11-20 15:08:27 : Done 24000 of 27870
2019-11-20 15:08:30 : Done 25000 of 27870
2019-11-20 15:08:33 : Done 26000 of 27870
2019-11-20 15:08:36 : Done 27000 of 27870

OG0026960_tree_id.txt - WARNING: ETE could not interpret tree file, it will be ignored
11323 trees had all species present and will be used by STAG to infer the species tree

Best outgroup(s) for species tree
---------------------------------
2019-11-20 15:43:28 : Starting STRIDE
2019-11-20 15:43:52 : Done STRIDE
Observed 325 well-supported, non-terminal duplications. 309 support the best root and 16 contradict it.
Best outgroup for species tree:
  triticeae_barley

Reconciling gene trees and species tree
---------------------------------------
Outgroup: triticeae_barley
2019-11-20 15:43:52 : Starting Recon and orthologues
2019-11-20 15:43:52 : Starting OF Orthologues
2019-11-20 15:43:56 : Done 0 of 27870
2019-11-20 15:45:39 : Done 1000 of 27870
2019-11-20 15:46:29 : Done 2000 of 27870
2019-11-20 15:47:09 : Done 3000 of 27870
2019-11-20 15:47:44 : Done 4000 of 27870
2019-11-20 15:48:11 : Done 5000 of 27870
2019-11-20 15:48:32 : Done 6000 of 27870
2019-11-20 15:48:52 : Done 7000 of 27870
2019-11-20 15:49:10 : Done 8000 of 27870
2019-11-20 15:49:28 : Done 9000 of 27870
2019-11-20 15:49:46 : Done 10000 of 27870
2019-11-20 15:50:02 : Done 11000 of 27870
2019-11-20 15:50:20 : Done 12000 of 27870
2019-11-20 15:50:38 : Done 13000 of 27870
2019-11-20 15:50:54 : Done 14000 of 27870
2019-11-20 15:51:11 : Done 15000 of 27870
2019-11-20 15:51:28 : Done 16000 of 27870
2019-11-20 15:51:42 : Done 17000 of 27870
2019-11-20 15:51:57 : Done 18000 of 27870
2019-11-20 15:52:12 : Done 19000 of 27870
2019-11-20 15:52:27 : Done 20000 of 27870
2019-11-20 15:52:40 : Done 21000 of 27870
2019-11-20 15:52:53 : Done 22000 of 27870
2019-11-20 15:53:05 : Done 23000 of 27870
2019-11-20 15:53:16 : Done 24000 of 27870
2019-11-20 15:53:26 : Done 25000 of 27870
2019-11-20 15:53:34 : Done 26000 of 27870
Traceback (most recent call last):
  File "orthofinder/orthofinder.py", line 1647, in <module>
  File "orthofinder/orthofinder.py", line 1482, in GetOrthologues
  File "orthofinder/scripts/orthologues.py", line 1002, in OrthologuesWorkflow
  File "orthofinder/scripts/orthologues.py", line 800, in ReconciliationAndOrthologues
  File "orthofinder/scripts/trees2ologs_of.py", line 644, in DoOrthologuesForOrthoFinder
TypeError: 'NoneType' object is not iterable
[63168] Failed to execute script orthofinder


```

**Comment by fei0810** [comment]
Maybe I find the reason, but I'm still a little confused.
In other issue, I find the MCL warning is normal, so I check the second warning about OG0026960_tree_id.txt. 
I find OG0026960_tree_id.txt is empty.  I also check the file OG0026960.phy in Distances directory.

```
cat OG0026960.phy
4
1_23413 0.500000 0.568843 0.625727 0.156944
3_19713 0.568843 0.500000 0.625727 0.489313
4_20549 0.625727 0.625727 inf 0.625727
7_23524 0.156944 0.489313 0.625727 0.500000
```

Then, I test FastME use OG0026960.phy as input. 

```
fastme -i OG0026960.phy -o out.tree -w O -s -n

 . This analysis will run on 112 threads.

#  Analysing dataset 1

 . Error: Invalid distance matrix : numerical value expected for taxon '4_20549' instead of 'inf'

```

Because `inf` value in OG0026960.phy , fastme can't work and  OG0026960_tree_id.txt is empty.  May be I could remove 4_20549 sequence from my input file then rerun orthofinder.

But I still don't know why the result is inf instead of 0.5.

**Comment by davidemms** [comment]
Hi

This issue has now been fixed, please see here: https://github.com/davidemms/OrthoFinder/issues/298

Thanks for providing the extra information about the failing gene tree, I will see if I can find out why it happened. Would you be able to tell me the sequence for 4_20549, you can find it in WorkingDirectory/Species4.fa.

Many thanks
David

**Comment by fei0810** [comment]
> Hi
> 
> This issue has now been fixed, please see here: #298
> 
> Thanks for providing the extra information about the failing gene tree, I will see if I can find out why it happened. Would you be able to tell me the sequence for 4_20549, you can find it in WorkingDirectory/Species4.fa.
> 
> Many thanks
> David

Hi

After deleting the 4_20549 sequence, I still encountered the same error. 

```
cat OG0026950.phy
4
1_23413 0.500000 0.570188 0.156944 0.627206
3_19713 0.570188 0.500000 0.490629 0.627206
7_23524 0.156944 0.490629 0.500000 0.627206
9_6643 0.627206 0.627206 0.627206 inf
```

The new sequence 9_6643 and the previous three sequences make up a new error cluster.

So I thought there might not be a problem with my sequence, and deleting it doesn't seem to be a reasonable solution.

Finally, I did three tests.

1. use the OrthoFinder V 2.3.3 with the default parameters

```
orthofinder -t 80 -p all_pep
```

2. use the OrthoFinder V 2.3.7 with -M msa parameter

```
orthofinder -t 80 -f all_pep -M msa
```

3. use the OrthoFinder V 2.3.7  with the default parameters also delete five 'wrong' sequences

```
orthofinder -t 80 -p ./ -f remove_bad_pep
```

None of the three tests went error, but I haven't carefully examined the differences between all the results. So I think the error mey be  come from dendroblast.

The sequences are shown below, 4_20549 and 9_6643 have the same protein sequence.

```
>4_20549
MSDPKYAYPYPAQGYYQGGPYQGQGGPYQGPPVMAPPQYAAPPPRAQPSFLEGCLAALCC
CCLIDECCCDPSIIFVG

>9_6643
MSDPKYAYPYPAQGYYQGGPYQGQGGPYQGPPVMAPPQYAAPPPRAQPSFLEGCLAALCC
CCLIDECCCDPSIIFVG
```

**Comment by davidemms** [comment]
I should have clarified, the fix for the issue 298 is from after the 2.3.7 release. See that issue page for details about using it. However, I'm working on getting a new release out today so you would ject be able to use that one instead.

I think the infinities might be caused by diamond not returning a self-self hit for these sequences. You could check this by looking in the WorkingDirectory/Blast4_4.txt.gz file to see if there is a self-self hit for 4_20549. If you look into this can you let me know the answer.

Assuming this is the problem, I will submit a fix that addresses that. 

All the best
David

**Comment by fei0810** [comment]
> I should have clarified, the fix for the issue 298 is from after the 2.3.7 release. See that issue page for details about using it. However, I'm working on getting a new release out today so you would ject be able to use that one instead.
> 
> I think the infinities might be caused by diamond not returning a self-self hit for these sequences. You could check this by looking in the WorkingDirectory/Blast4_4.txt.gz file to see if there is a self-self hit for 4_20549. If you look into this can you let me know the answer.
> 
> Assuming this is the problem, I will submit a fix that addresses that.
> 
> All the best
> David

Hi, 

Because I have delated the 4_20549 sequence,  I checked another inf sequence 9_6643 in the WorkingDirectory/Blast9_9.txt.gz ,  there is no self-self hit like other sequence.

**Comment by davidemms** [comment]
Thanks, the submitted change should fix this. It is available in the new release v2.3.8.

All the best
David
